internal/codegen/tpl_module_test.go

```diff
@@ -0,0 +1,159 @@
+package codegen
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"go.rgst.io/stencil/internal/modules"
+	"go.rgst.io/stencil/internal/modules/modulestest"
+	"go.rgst.io/stencil/pkg/configuration"
+	"go.rgst.io/stencil/pkg/slogext"
+	"gotest.tools/v3/assert"
+)
+
+// TestTplModule_Tpl tests [TplModule] in a template context.
+func TestTplModule_Tpl(t *testing.T) {
+	tests := []struct {
+		name             string
+		functionTemplate string
+		callingTemplate  string
+		want             string
+		wantErrContains  string
+	}{
+		{
+			name:            "should error on non-existent template",
+			callingTemplate: `{{- module.Call "caller.HelloWorld" }}`,
+			wantErrContains: `calling Call: module "caller" did not register any functions or was not imported`,
+		},
+		{
+			name: "should support calling exported function",
+			functionTemplate: `{{- define "HelloWorld" -}}
+{{ return "Hello, world!" }}
+{{- end -}}
+{{- module.Export "HelloWorld" -}}`,
+			callingTemplate: `{{ module.Call "function.HelloWorld" }}`,
+			want:            "Hello, world!",
+		},
+		{
+			name: "should support returning errors",
+			functionTemplate: `{{- define "HelloWorld" -}}
+{{ return "" (error "Failed!") }}
+{{- end -}}
+{{- module.Export "HelloWorld" -}}`,
+			callingTemplate: `{{ module.Call "function.HelloWorld" }}`,
+			wantErrContains: "Failed!",
+		},
+		{
+			name: "should not continue processing after error",
+			functionTemplate: `{{- define "HelloWorld" -}}
+{{ return "" (error "Failed!") }}
+{{ fail "should not continue processing" }}
+{{- end -}}
+{{- module.Export "HelloWorld" -}}`,
+			callingTemplate: `{{ module.Call "function.HelloWorld" }}`,
+			wantErrContains: "Failed!",
+		},
+		{
+			name: "should support passing concrete types",
+			functionTemplate: `{{- define "HelloWorld" -}}
+{{ return (fromYaml "hello: world") }}
+{{- end -}}
+{{- module.Export "HelloWorld" -}}`,
+			callingTemplate: `{{ (module.Call "function.HelloWorld").hello }}`,
+			want:            "world",
+		},
+		{
+			name: "should pass data between calls",
+			functionTemplate: `{{- define "HelloWorld" -}}
+{{ return .Data }}
+{{- end -}}
+{{- module.Export "HelloWorld" -}}`,
+			callingTemplate: `{{ module.Call "function.HelloWorld" "hello" }}`,
+			want:            "hello",
+		},
+		{
+			name: "should use data from caller",
+			functionTemplate: `{{- define "HelloWorld" -}}
+{{ return (file.Path) }}
+{{- end -}}
+{{- module.Export "HelloWorld" -}}`,
+			callingTemplate: `{{ module.Call "function.HelloWorld" }}`,
```

Comment:
add fail case for calling "blah"

Reply:
Non-existent? That's the first test case

Reply:
oic